README.md
[![Build Status](https://travis-ci.org/jaywink/social-relay.svg?branch=master)](https://travis-ci.org/jaywink/social-relay) [![codecov](https://codecov.io/gh/jaywink/social-relay/branch/master/graph/badge.svg)](https://codecov.io/gh/jaywink/social-relay) [![Requirements Status](https://requires.io/github/jaywink/social-relay/requirements.svg?branch=master)](https://requires.io/github/jaywink/social-relay/requirements/?branch=master) 

[![issue tracker](https://img.shields.io/badge/issue%20tracker-gitlab-orange.svg)](https://git.feneas.org/jaywink/social-relay)

# Social-Relay

Application to act as a relay for public posts using the Diaspora protocol. Keeps track of nodes and their subscription preferences, receives payloads and forwards the payloads to subscribers. The aim is to pass public posts around in an efficient way so any new node in the network can quickly subscribe to lots of public activity, without having to wait a long time to create social relationships.

## How does one integrate to the relay system? How do I write my own relay?

See [relay design concept](https://github.com/jaywink/social-relay/blob/master/docs/relays.md).

Original idea for the relay system can be found in the [diaspora* project wiki](https://wiki.diasporafoundation.org/Relay_servers_for_public_posts).

## Installation
       
### System libraries

Depending on your database, you will probably need extra libraries installed.

For databases, PostgreSQL and MySQL/MariaDB are supported, choose one.

#### Python

Python 3.4+ is required.

#### Redis

    # Debian / Ubuntu
    sudo apt-get install redis-server

#### PostgreSQL

    # Debian / Ubuntu
    sudo apt-get install libpq-dev postgresql
    
#### MySQL/MariaDB

    # Debian / Ubuntu
    sudo apt-get install python3-dev libmysqlclient-dev 

    # Red Hat / CentOS
    sudo yum install python3-devel mysql-devel
    
#### Federation

The `federation` dependency `lxml` requires certain libraries present:
 
    # Debian / Ubuntu
    sudo apt-get install libxml2-dev libxslt-dev lib32z1-dev python3-dev

### Python libraries

Create a Python 3.4+ virtualenv and activate it.

Ensure `pip` and `setuptools` are up to date.

    pip install -U pip setuptools

For a production deployment, choose your database and install the requirements:

    # PostgreSQL
    pip install -r requirements/postgresql.txt
    
    # MySQL/MariaDB
    pip install -r requirements/mysql.txt

### Configuring

Create local config:

    cp social_relay/local_config.py.example social_relay/local_config.py

Edit the `local_config.py` file as instructed in the file.

### Database

You can set database connection settings as environment variables or in `social_relay/local_config.py` - see `social_relay/config.py` for what to override.

The database needs an initial schema creation. Do this with:

    arnold up 0

The same command should always be run when fetching new relay code. It will migrate any new schema changes.

### RQ Dashboard

An RQ dashboard can be found at `/rq`. Enable it in `social_relay/local_config.py` by setting `RQ_DASHBOARD = True`.
You must also set a username and password in the same file.

### Static files

Bower is used to pull in some JavaScript libs. [Install it first](http://bower.io/) if needed. Then run `bower install`.

Statics are server under the `/static` path which should be server by the web server.

### Running tasks and workers

Scheduled jobs handle the polling of node lists and nodes themselves, to fetch their subscription settings. Without the scheduled jobs the server will not be able to function. RQ workers on the other hand process all the incoming payloads and distribute them onwards to subscribing nodes. At least one RQ worker must be running at all times.

In production, it's easiest to use the provided `circus` configuration. This is installed via the provided production requirements, or `pip install circus` if not using the provided requirements file.

Then, export how many RQ workers you want. If you see your `receive` queue build up, increase this count and restart `circus`.

    export RQWORKER_NUM=5

To start `circus`, virtualenv activated in the project folder:

    circusd extras/circus/circus.ini
    
You can daemonize `circus` by passing an extra `--daemonize` flag.

#### Running tasks manually (without circus)

If you don't want to use `circus`, run the tasks manually.  Keep this running:

    python -m tasks.schedule_jobs

#### Processing receive queue (without circus)

If you don't want to use `circus`, run the workers manually. Run the worker(s) as follows:

    rqworker -c social_relay.config receive

## Deployment

Pretty much normal Python + WSGI setup, just install the requirements and serve app using WSGI and statics via the web server. See the following sections for platform specific helpers.

An Apache2 site example can be found [here](https://github.com/jaywink/social-relay/blob/master/ansible/roles/social-relay/templates/apache.conf.j2). The same folder also has examples for upstart init jobs.

### Ansible (Ubuntu)

An Ansible role written for Ubuntu is provided in `ansible` directory. The role uses PostgreSQL, uWSGI and Apache. It will run also the scheduled jobs and a worker. Everything is handled by upstart.

Tested with Ubuntu 14.04 LTS.

### SystemD service files

There are example systemd service files in the 'extra' directory. The examples
use a specific user and utilize gunicorn. They have been tested on CentOS 7.

To use, modify as needed (user, group, and path), copy to `/etc/systemd/service`
and start/enable as such:

    systemctl start social-relay_server.service
    systemctl start social-relay_tasks.service
    systemctl start social-relay_rqworker@receive.service
    systemctl start social-relay.target

    systemctl enable social-relay_server.service
    systemctl enable social-relay_tasks.service
    systemctl enable social-relay_rqworker@receive.service
    systemctl enable social-relay.target

The rqworker service file can also be used to start the optional `failed` queue as well.

    systemctl start social-relay_rqworker@failed.service

## Development

### Development requirements

    pip install -r requirements/development.txt
    
Ensure also `bower` is installed.

    npm install bower
    bower install

### Running a development server

This is not the recommended way for a production server. For testing and development, run the server:

    python devserver.py
    
The app will be running at [http://127.0.0.1:5000](http://127.0.0.1:5000).

### Running an RQ worker

    rqworker -c social_relay.config receive

### Running tests

Ensure there is a test database. By default the tests try to connect to PostgreSQL with db, username and password `socialrelaytest`.

Then run test with:

    py.test

### Maintaining dependencies

When changing requirements, use pip-tools.

    pip install pip-tools
    
To compile all

    ./compile-requirements.sh

## Author

Jason Robinson / @jaywink / https://jasonrobinson.me

Awesome contributors listed in `CONTRIBUTORS` file.

## License

AGPLv3


CHANGELOG.md
## [unrelased]

### Changed

* Now using `pip-tools` to maintain dependencies. To install all requirements now just choose a db and install either `requirements/postgresql.txt` or `requirements/mysql.txt` which will pull in all the rest.

* Removed tracking of where posts are sent. This was used to allow sending participations (likes, comments and retractions) to the same targets. Diaspora project rejected this enhancement from their code and it ended up only being used by Hubzilla. Diaspora now sends a participation automatically on receiving any post which should be the default behaviour. Thus, no need to relay participations.

## [1.5.0] - 2017-09-25

### Changed
* Start sending payloads according to new style Diaspora protocol requirements. #[62](https://github.com/jaywink/social-relay/issues/62).

## [1.4.0] - 2017-07-23

### Added
* Accept new style Diaspora public payloads without `xml=payload` form data.
* Add profile model. Store remote profile handle + public key for later use, since we don't want to always fetch them.
* Start validating signatures in sent payloads. This requires fetching remote profiles. Closes #[31](https://github.com/jaywink/social-relay/issues/31).

## [1.3.3] - 2017-03-10

### Fixed
* Fix sending to nodes via tags when tags in content was not all lower case.

## [1.3.2] - 2017-03-06

### Fixed
* Add a longer worker timeout for sending. Seems default 180 seconds is not enough in cases where post will be delivered to a lot of hosts. Allow setting a long timeout, defaulting to 1080 seconds.
* Refactor sending to not send payload multiple times in the case that multiple entities are found in the payload. In reality this should never happen, since Diaspora payloads have one top level entity only.

## [1.3.1] - 2017-01-28

### Fixed
* Diaspora changed their timestamp format in release 0.6.3. This release brings federation support for this timestamp format. Any relay that wants to relay content from Diaspora upgraded Diaspora nodes must upgrade to this release asap.

### Added
* Tests are now built also against Python 3.6.

## [1.3.0] - 2017-01-11

### Backwards incompatible changes

#### Database engines

**Important!** Removed support for SQLite. Added support for PostgreSQL and MySQL/MariaDB.

Reason for this is that SQLite simply will not scale for more than one worker running. With increased adoption of the relay system within The Federation, relays will have to deal with more traffic in the future. Thus, a proper database is needed.

Relays upgrading should not worry too much about migrating the old data over. That can be done but due to limited amount of nodes sending participations at this moment, routing doesn't yet use the database that much. There are no schema changes in this upgrade so if you wish to move the data from the SQLite DB, it should be relatively easy to do so.

#### RQ worker wrapper

The old wrapper `runworker.py` has been removed. This was initially added to easily provide a way to run a worker using the already existing Redis configuration. It's better however to use the cli command provided by RQ to not have to maintain our own wrapper, and to be able to use all the available cli options.

To run a worker, use the following command:

    rqworker -c social_relay.config receive
    
Note! See `circus` note below for recommended way of running workers in production.

### Added
- Added `circus` to handle running jobs and and one or more workers comfortably via one command. This removes the need to separately ensure jobs and workers are running. To use `circus`, either install it via `requirements/production.txt` or separately with `pip install circus`. To run jobs and workers, use the following command: `RQWORKER_NUM=5 circusd extras/circus/circus.ini` - specifying in the variable the amount of workers you want running.

## [1.2.0] - 2016-10-25

### Added
- Expose [NodeInfo](https://github.com/jhass/nodeinfo) to allow registering relays to pod lists. Unfortunately, NodeInfo schema doesn't contain the relay software key so this NodeInfo document cannot be validated by consumers.
- Network calls now use a custom user agent `Social-Relay/<version> - https://github.com/jaywink/social-relay`. Thanks @bmanojlovic for the patch.
- Relay also Diaspora `retraction` and `photo` entities. The former follows the same way `like` and `comment` entities are relayed, ie the targets are the same as where the `status_message` or `photo` were relayed. `photo` follows same rules as `status_message` entities, ie according to subscriber wishes.

### Changed
- Replaced custom payload sending with the new helper from `federation`. This will not try to deliver to `http` targets at all. This means nodes that are not using https will not get deliveries. Really, these days, there is no reason to run a public website with http.

### Removed
- Removed suggestion to use `pip-tools` and convert requirements files to standard Python project requirements files. Didn't dig the workflow after all. To install dev dependencies use `requirements/development.txt`. For production, use `requirements/production.txt`, which also contains `uWSGI`. If you don't deploy using uWSGI, you can just use `requirements/requirements.txt`.

## [1.1.1] - 2016-05-09

### Changed
- Bump `Social-Federation` to 0.3.2. Fixes issue with DiasporaPost processing when actual message content is empty.

## [1.1.0] - 2016-05-08

### Backwards incompatible changes

Python 2.x support has been dropped. The current tested Python versions are 3.5 and 3.6.

### Important!

New dependency, SQLite3 added. Make sure before deploying to install necessary OS packages (`sqlite3` in Ubuntu).

Of course a database means schema migrations. See the readme for instructions on how to create the initial schema and what to do when pulling in a new version of the relay.

### Python dependencies


social_relay/config.py
# No need to change this normally, override in locals if needed
import logging
from logging import handlers
import os

from peewee import PostgresqlDatabase, MySQLDatabase

# Current version. Needless to say this shouldn't be changed except for a release.
VERSION = "1.6.0-dev"

# Set this to a real domain in local config
SERVER_HOST = "http://127.0.0.1:5000"

RELAY_USERNAME = "relay"


# Just something hcard needs, override in locals if you want to customize
# This might appear visible on some pod who fetches the relay hcard...
RELAY_NAME = "Social-Relay"

# Set this to something else if you have any issues from default value
USER_AGENT = "Social-Relay/%s - https://github.com/jaywink/social-relay" % VERSION

# Relays only need a public key because Diaspora pods want to read one. So this doesn't need changing normally,
# override in locals if needed
RELAY_PUBLIC_KEY = "ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQC9eofav26VCq3C6g9sIPjiEYzhuiFGGAMx4NAY2bP9NFxXtuERYsqtS+" \
                   "bie101/rjcDEbK5IcZY+jvCGh0P8oI/nj5kSREWxNv3OwRlCbyCtN2XR/eWEs/SfNvZysGtQunLGaUwtcyD+PUXeDuj0PH" \
                   "FuVUEudJ5DiJzu2fcUMGhLv9Xr+C2HU0tcNUjtZmYySsiVXVVrKMzj/Bgefkq7ECy3gchSgYpa2Mc5vOEBG1OxAyQvYDlE" \
                   "IohseDPo5YC9FBjp9gy7caJtvSmM5YQnDP7fjp/ghDSFtVqRWpfJbM2g9R+rAhD6w1JgvBYoOHF0ittOrbPv8WvzhaHivX" \
                   "ZADp relay@localhost"

POD_LIST_JSON = "https://the-federation.info/pods.json"

# RQ Dashboard
RQ_DASHBOARD = False  # Enable in local config
RQ_POLL_INTERVAL = 2500  # Poll interval for the RQ dashboard

BOWER_COMPONENTS_ROOT = "../bower_components"

LOG_PATH = "var/social-relay.log"
LOG_TO_CONSOLE = False

# Worker RQ timeout (seconds)
# Sending to many hosts can take some time. If you see timeout errors in the RQ dashboard, set a longer timeout here
RELAY_WORKER_TIMEOUT = 1080

# Database
# Override in local_config.py or env variable "RELAY_DATABASE_TYPE"
DATABASE_TYPE = os.environ.get("RELAY_DATABASE_TYPE") or 'postgresql'
DATABASE_NAME = os.environ.get("RELAY_DATABASE_NAME") or 'socialrelay'
DATABASE_USER = os.environ.get("RELAY_DATABASE_USER") or 'socialrelay'
DATABASE_PASSWORD = os.environ.get("RELAY_DATABASE_PASSWORD")
DATABASE_HOST = os.environ.get("RELAY_DATABASE_HOST") or '127.0.0.1'

# Redis
REDIS_DB = 0
REDIS_PORT = 6379
REDIS_HOST = 'localhost'

# Override in local config with hosts to always forward to
ALWAYS_FORWARD_TO_HOSTS = []

try:
    from social_relay.local_config import *
except ImportError:
    pass

RELAY_ACCOUNT = "%s@%s" % (
    RELAY_USERNAME,
    SERVER_HOST.split("//")[1]
)

# If running tests load some overrides here
if "TEST" in os.environ:
    from social_relay.tests.config import *

# Logging init
file_handler = handlers.RotatingFileHandler(
    filename=os.path.join(os.path.dirname(os.path.abspath(__file__)), "..", LOG_PATH),
    maxBytes=10000000,
    backupCount=10
)
logging_handlers = [file_handler]
if LOG_TO_CONSOLE:
    logging_handlers.append(logging.StreamHandler())

logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s:%(levelname)s:%(module)s: %(message)s',
    handlers=logging_handlers
)

# Make sure we have a GUID
try:
    assert len(RELAY_GUID) == 32
except Exception:
    print("****\nYOU MUST define 'RELAY_GUID' configuration as 32 chars that will be unique *across* the "
          "whole network.\n****")
    raise

# Make sure if RQ_DASHBOARD is enabled, we have user/pass
if RQ_DASHBOARD:
    try:
        assert RQ_DASHBOARD_USERNAME
        assert RQ_DASHBOARD_PASSWORD
    except Exception:
        print("****\nYOU MUST define RQ_DASHBOARD_USERNAME and RQ_DASHBOARD_PASSWORD if RQ_DASHBOARD is enabled.\n"
              "****")
        raise

# Initialize db connection
if DATABASE_TYPE == "postgresql":
    database = PostgresqlDatabase(
        database=DATABASE_NAME, user=DATABASE_USER, password=DATABASE_PASSWORD, host=DATABASE_HOST,
        autorollback=True,
    )
elif DATABASE_TYPE == "mysql":
    database = MySQLDatabase(
        database=DATABASE_NAME, user=DATABASE_USER, password=DATABASE_PASSWORD, host=DATABASE_HOST,
    )
else:
    raise Exception("Invalid database type configuration")
